turqstat.1
.TH TURQSTAT 1 2008 "Softwolves Software" ""
.SH NAME
turqstat \- Statistics tool for Fidonet and Usenet
.SH SYNOPSIS
.PD 0
.B turqstat
[options] outputfile areadef(s)
.PD
.SH DESCRIPTION
.B Turquoise SuperStat
reads the contents of a Fidonet message area in one of its supported
formats, or a standard Usenet news spool, and creates statistics about this:
.RS
.PP
.I Quoter blacklist.
.PD 0
.PP
.I Toplist of senders.
.PP
.I Toplist of original content per message.
.PP
.I Toplist of Fidonet nets (Fidonet only).
.PP
.I Toplist of Internet topdomains.
.PP
.I Toplist of receivers (Fidonet only).
.PP
.I Toplist of subjects.
.PP
.I Toplist of used software.
.PP
.I Postings by weekday.
.PP
.I Postings by time of day.
.RE
.PD
.SH OPTIONS
.SS "DATA SELECTION"
.TP
.I \-d days
Selects the number of days back to count messages from. The default is
0, which mean count all messages.
.TP
.I \-n num
Selects the maximum number of entries in the toplists. The default is 15.
0 indicates no limit.
.TP
.I \-r range
Selects a range of dates and times for which statistics will be gathered.
The range is given on the form
.RI [ yyyymmdd [T hhmmss ]]\-[ yyyymmdd [T hhmmss ]].
Examples:
.B 20000101T000000\-20001231T235959
is all of 2000,
.B \-20010101
is everything before (not including) January 1st, 2001, and
.B 20010101\-
is everything from (and including) January 1st, 2001.
.SS "OUTPUT FORMAT"
.TP
.I \-F file
Selects the template file to use. The default is
.B default.tpl
.SS "INPUT SELECTION"
.TP
.I \-a path/server
Selects the message base path for the FDAPX/w, MyPoint and taanstafl
message bases, which have all message bases in the same files.
Sects NNTP server for non-local Usenet news (\-U).
.TP
.I \-s
Squish message bases (default).
.TP
.I \-m
FTSC *.MSG message bases.
.TP
.I \-o
OPUS *.MSG message bases.
.TP
.I \-j
JAM message bases.
.TP
.I \-f
FDAPX/w message bases.
.TP
.I \-t
Tanstaafl message bases.
.TP
.I \-p
MyPoint message bases.
.TP
.I \-u
Usenet news spool (local netnews).
.TP
.I \-U
Usenet news (NNTP) server (remote netnews).
.SS "TOPLIST SELECTION"
These options are used to turn toplists off.
By default, all toplists are enabled (except for Internet top domains, see
below).
.TP
.I \-Q
Turns the quoter blacklist off.
.TP
.I \-W
Turns the writer toplist off.
.TP
.I \-O
Turns the original content per message toplist off.
.TP
.I \-N
Turns the top nets (Fidonet only) list off.
.TP
.I \-R
Turns the recipient (Fidonet only) toplist off.
.TP
.I \-T
Turns the top Internet topdomain list on/off. For Usenet, the default is
on, and this switch turns it off. For Fidonet, the default is off, and this
switch turns it on, but only if there are any RFC gated messages detected.
.TP
.I \-S
Turns the subject toplist off.
.TP
.I \-P
Turns the program toplist off.
.TP
.I \-H
Turns the hourly posting statistics off.
.TP
.I \-D
Turns the daily posting statistics off.
.TP
.I \-V
Turns the display of versions in the program toplist off.
.SS "OUTPUT OPTIONS"
.TP
.I \-A
Display all numbers in toplists (default is off).
.TP
.I \-C charset
Selects character encoding to use for the created report files.
The character encoding name can be specified by either its Fidonet or MIME
name.
Default for the Win32 and OS/2 versions is IBMPC (codepage 437), and
for all other versions LATIN-1 (iso-8859-1).
.TP
.I \-L
Use locale defined date format in output (default is ISO-8601 format, as
per Swedish and EU standards).
.SH "SEE ALSO"
.BR xturqstat (1),
.I turqstat.doc
.SH BUGS
Report any bugs to the authors.
.SH AUTHOR
Turquoise SuperStat was written by Peter Krefting.
If you have bug reports or questions, mail them to
peter@softwolves.pp.se or see the website at
http://www.softwolves.pp.se/sw/
.PP
A Softwolves Software (tm) Release in 2008.
.PP
This program is released under the GNU General Public License, version 2.


ChangeLog
Version 3.1 - UNRELEASED

 * Added HTML output support.

Version 3.0 - 2008-01-21

 * Added support for templates.
 * Changed to use Visual C++ Express for the Windows version.
 * Updates for GCC 4.


turqstat.cpp
// Turquoise SuperStat
//
// A statistic collection program for Fidonet and Usenet systems
//
// Copyright (c) 1998-2008 Peter Krefting
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 2
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

#include <config.h>
#ifdef HAVE_IOSTREAM
# include <iostream>
#else
# include <iostream.h>
#endif
#include <string>
#ifdef HAVE_GETOPT_IN_UNISTD
# include <unistd.h>
#elif defined(HAVE_GETOPT_IN_STDIO)
# include <stdio.h>
#elif defined(HAVE_GETOPT_IN_GETOPT)
# include <getopt.h>
#endif
#include <stdlib.h>
#include <time.h>
#ifdef HAVE_LOCALE_H
# include <locale.h>
#endif
#ifdef HAVE_WINSOCK2_H
# include <winsock2.h>
#endif
#ifdef HAVE_WIN32_GETMODULEFILENAME
# include <windows.h>
#endif
#include <string.h>

#include "statengine.h"
#include "statview.h"
#include "arearead.h"
#include "squishread.h"
#include "fdapxread.h"
#include "jamread.h"
#include "mypointread.h"
#include "sdmread.h"
#include "tanstaaflread.h"
#include "newsspoolread.h"
#if defined(HAVE_NNTP)
# include "nntpread.h"
#endif
#include "version.h"
#include "utility.h"
#include "mytime.h"
#include "output.h"
#include "template.h"

/**
 * Wrapper class to retrieve data. This class is called from the main
 * procedure to retrieve data from the specified message bases into the
 * engine, and to connect that engine to the view to create a report file.
 */
class StatRetr
{
public:
    /**
     * Numeric constants describing the message base types that are
     * supported by Turquoise SuperStat.
     */
    enum basetype_e
    {
        squish,     ///< SquishMail message base format.
        sdm,        ///< Standard FTSC message base format.
        opus,       ///< Opus SDM message base format.
        fdapx,      ///< FDAPX/w message base format.
        jam,        ///< JAM message base format.
        mypoint,    ///< MyPoint message base format.
        tanstaafl,  ///< Tanstaafl message base format.
#if defined(HAVE_NNTP)
        nntp,       ///< Usenet news server (nntp).
#endif
        newsspool   ///< Usenet news spool.
    };

    /**
     * Constructor that does the work.
     * @param engine         Reference to engine that stores data.
     * @param view           Reference to view that creates report.
     * @param areapath       Paths to areas to be examined.
     * @param numpaths       Number of areas in array.
     * @param outputfilepath Path to output file to create.
     * @param basepath       Base path of message base (if required).
     * @param basetype       Message base type for input.
     * @param rangestart     First date to save data for.
     * @param rangeend       Last date to save data for.
     */
    StatRetr(StatEngine &, StatView &,
             char **areapath, int numpaths, char *outputfilepath,
             char *basepath, basetype_e basetype,
             time_t rangestart, time_t rangeend);

    /** Standard destructor. */
    ~StatRetr();
};

/** Display the program's help screen. */
void helpscreen(void);

/** Evaluate a date-time range as given on the command line.
 * @param rangespec      Range specification string.
 * @param range_start_pp Where to create the range start time.
 * @param range_end_pp   Where to create the range end time.
 */
void evaluaterange(const char *rangespec,
                   time_t **range_start_pp, time_t **range_end_pp);

int main(int argc, char *argv[])
{
    unsigned days = 0;
    unsigned maxnum = 15;
    char *basepath = NULL;
	const char *templatefile = NULL;
    StatRetr::basetype_e basetype = StatRetr::squish;
    bool quoters = true, topwritten = true, topreceived = true,
         topsubjects = true, topprograms = true, weekstats = true,
         daystats = true, versions = true, allnums = false,
         toporiginal = true, topnets = true, topdomains = true;
    time_t *range_start_p = NULL, *range_end_p = NULL;
#if defined(__EMX__) || defined(__CYGWIN__) || defined (__MINGW32__)
    const char *charset = "IBMPC";
#else
    const char *charset = "LATIN-1";
#endif
#if defined(HAVE_LOCALE_H) || defined(HAVE_OS2_COUNTRYINFO) || defined(HAVE_WIN32_LOCALEINFO)
    bool uselocale = false;
#endif
#if defined(HAVE_WIN32_GETMODULEFILENAME)
	char defaulttemplate[_MAX_PATH];
#endif

    // Setup locale
#ifdef HAVE_LOCALE_H
    setlocale(LC_TIME, "");
#endif

    // Display banner
    cout << "Turquoise " << version
         << " - Statistics tool for Fidonet and Usenet message bases"
         << endl;
    cout << copyright << ". GNU GPL 2 licensed." << endl;
    cout << "A Softwolves Software Release in " << releaseyear << "." << endl;
    cout << endl;

    // Handle arguments
    int c;
#if defined(HAVE_LOCALE_H) || defined(HAVE_OS2_COUNTRYINFO) || defined(HAVE_WIN32_LOCALEINFO)
    while (EOF != (c = getopt(argc, argv, "d:n:a:r:C:F:smofjptuQWRSPOHDVNTALU?")))
#else
    while (EOF != (c = getopt(argc, argv, "d:n:a:r:C:F:smofjptuQWRSPOHDVNTAU?")))
#endif
    {
        switch (c)
        {
            case 'd':   days = strtoul(optarg, NULL, 10);           break;
            case 'n':   maxnum = strtoul(optarg, NULL, 10);         break;
            case 'a':   basepath = optarg;                          break;
            case 'r':   evaluaterange(optarg, &range_start_p, &range_end_p);
                                                                    break;
            case 'F':   templatefile = optarg;                      break;

            case 's':   basetype = StatRetr::squish;                break;
            case 'm':   basetype = StatRetr::sdm;                   break;
            case 'o':   basetype = StatRetr::opus;                  break;
            case 'f':   basetype = StatRetr::fdapx;                 break;
            case 'j':   basetype = StatRetr::jam;                   break;
            case 'p':   basetype = StatRetr::mypoint;               break;
            case 't':   basetype = StatRetr::tanstaafl;             break;
#if defined(HAVE_NNTP)
            case 'U':   basetype = StatRetr::nntp;                  break;
#endif
            case 'u':   basetype = StatRetr::newsspool;             break;

            case 'Q':   quoters = false;                            break;
            case 'W':   topwritten = false;                         break;
            case 'R':   topreceived = false;                        break;
            case 'S':   topsubjects = false;                        break;
            case 'P':   topprograms = false;                        break;
            case 'O':   toporiginal = false;                        break;
            case 'N':   topnets = false;                            break;
            case 'T':   topdomains = false;                         break;
            case 'H':   daystats = false;                           break;
            case 'D':   weekstats = false;                          break;

            case 'V':   versions = false;                           break;
            case 'A':   allnums = true;                             break;

#if defined(HAVE_LOCALE_H) || defined(HAVE_OS2_COUNTRYINFO) || defined(HAVE_WIN32_LOCALEINFO)
            case 'L':   uselocale = true;                           break;
#endif

            case 'C':   charset = optarg;                           break;

            case '?':
            default:
                helpscreen();
                return 1;
        }
    }

    if (argc - optind < 2)
    {
        cerr << "Illegal number of arguments (-? for help)." << endl;
        return 1;
    }

    // Check parameters for integrity
    if ((StatRetr::tanstaafl == basetype ||
         StatRetr::fdapx == basetype ||
         StatRetr::mypoint == basetype) &&
        !basepath)
    {
        cerr << "Base message path missing (-? for help)." << endl;
        return 1;
    }

    // Load template
    if (!templatefile)
    {
        // Use default template if none was supplied.
#if defined(HAVE_WIN32_GETMODULEFILENAME)
		// Windows: Template is in the same directory as the EXE file.
		GetModuleFileNameA(NULL, defaulttemplate, _MAX_PATH);
		char *p = strrchr(defaulttemplate, '\\');
		if (p && p < defaulttemplate + _MAX_PATH - 13)
		{
			strcpy(p + 1, "default.tpl");
			templatefile = defaulttemplate;
		}
#else
		// Unix-like environment: Template is in a pre-defined installation
		// directory.
        templatefile = DATA
#ifdef BACKSLASH_PATHS
            "\\"
#else
            "/"
#endif
            "default.tpl";
#endif
    }

    bool is_error = false;
    Template *output_template_p = Template::Parse(templatefile, is_error);
    if (!output_template_p || is_error)
    {
		TDisplay *display = TDisplay::GetOutputObject();
		display->ErrorQuit(TDisplay::template_parse_error, 2);
    }

    // Create engine object
    StatEngine engine;

    // Create view object
    StatView view;

    // Select output format
    view.SetTemplate(output_template_p);

    // Enable toplists we want
    view.EnableQuoters(quoters);
    view.EnableTopWritten(topwritten);
    view.EnableTopOriginal(toporiginal);
    view.EnableTopNets(topnets);
    if (StatRetr::newsspool == basetype)
        view.EnableTopDomains(topdomains);
    else
        view.EnableTopDomains(!topdomains);
    view.EnableTopReceived(topreceived);
    view.EnableTopSubjects(topsubjects);
    view.EnableTopPrograms(topprograms);
    view.EnableWeekStats(weekstats);
    view.EnableDayStats(daystats);
#if defined(HAVE_LOCALE_H) || defined(HAVE_OS2_COUNTRYINFO) || defined(HAVE_WIN32_LOCALEINFO)
    view.UseLocale(uselocale);
#endif

    // Set toplist parameters
    view.ShowVersions(versions);
    view.ShowAllNums(allnums);
    view.SetMaxEntries(maxnum);
    view.SetCharset(charset);

    // Compute starting time if days parameter is given
    if (days != 0 && !range_start_p)
    {
        range_start_p = new time_t(time(NULL) - days * 86400L);
        struct tm *fromtm = localtime(range_start_p);
        fromtm->tm_hour = 0;
        fromtm->tm_min  = 0;
        fromtm->tm_sec  = 0;
        *range_start_p = my_mktime(fromtm);
    }

    // Create retrieval object, and do the work
    StatRetr s(engine, view,
               &argv[optind + 1], argc - optind - 1, argv[optind],
               basepath, basetype, range_start_p ? *range_start_p : time_t(0),
               range_end_p ? *range_end_p : time_t(DISTANT_FUTURE));

    delete range_start_p;
    delete range_end_p;
    delete output_template_p;

    return 0;
}

StatRetr::StatRetr(StatEngine &engine, StatView &view,
                   char **areapath, int numpaths, char *outputfilepath,
                   char *basepath, basetype_e basetype,
                   time_t rangestart, time_t rangeend)
{
    unsigned areanum;

    // Transfer from area(s) to engine
    AreaRead *area;
    int counter = 0;
    while (numpaths)
    {
        area = NULL;
        switch (basetype)
        {
            case squish:
                area = new SquishRead(areapath[counter]);
                break;

            case sdm:
            case opus:
                area = new SdmRead(areapath[counter], basetype == opus);
                break;

            case fdapx:
                areanum = strtoul(areapath[counter], NULL, 10);
                area = new FdApxRead(basepath, areanum);
                break;

            case jam:
                area = new JamRead(areapath[counter]);
                break;

            case mypoint:
                areanum = strtoul(areapath[counter], NULL, 10);
                area = new MyPointRead(basepath, areanum);
                break;

            case tanstaafl:
                areanum = strtoul(areapath[counter], NULL, 10);
                area = new TanstaaflRead(basepath, areanum);
                break;

            case newsspool:
                area = new NewsSpoolRead(areapath[counter]);
                break;

#if defined(HAVE_NNTP)
            case nntp:
                area = new NntpRead(basepath, areapath[counter]);
                break;
#endif

            default:
                TDisplay::GetOutputObject()->
                    InternalErrorQuit(TDisplay::message_base_mismatch, 1);
                exit(1);
        }

        if (!area)
        {
            TDisplay::GetOutputObject()->ErrorQuit(TDisplay::out_of_memory_area,
                                                   1);
        }
        if (!(area->Transfer(rangestart, rangeend, engine))) return;

        engine.AreaDone();
        cout << "Finished reading " << areapath[counter] << endl;

        delete area;
        counter ++;
        numpaths --;
    }

    // Write output
    view.CreateReport(&engine, outputfilepath);
    cout << "Wrote " << outputfilepath << endl;
}

StatRetr::~StatRetr()
{
    cout << "Turquoise finished." << endl;
}

void helpscreen(void)
{
    cout << "Usage: turqstat [options] outputfile areadef(s)" << endl;
    cout << endl;
    cout << "Data selection and display options:" << endl;
    cout << "  -a path  Base path (see below)    -d days   Days back to count"
         << endl;
    cout << "  -n num   Maximum entries          -r range  Date range"
         << endl;
    cout << "  -F file  Use template file (default: default.tpl)" << endl;
    cout << endl;
    cout << "Input selection options:" << endl;
    cout << "  -s  Squish       -j  JAM          -p  MyPoint*" << endl;
    cout << "  -m  FTSC *.MSG   -f  FDAPX/w*"
#if defined(HAVE_NNTP)
                                            "     -U  Usenet nntp server**"
#endif
         << endl;
    cout << "  -o  Opus *.MSG   -t  Tanstaafl*   -u  Usenet news spool" << endl;
    cout << "    * = requires -a parameter         ** = use -a for server" << endl;
    cout << endl;
    cout << "Output selection options (turns off respective list):" << endl;
    cout << "  -Q  Quoters     -S  Subjects   -N  Fidonet nets   -H Hour stats"
         << endl;
    cout << "  -W  Writers     -P  Programs   -T  Topdomains*  "
            "  -D Weekday stats" << endl;
    cout << "  -R  Receivers   -O  Original       * = off for Usenet, on for "
            "Fidonet" << endl;
    cout << "  -V          No version info    -A  Show all numbers in toplists"
         << endl;
#if defined(HAVE_LOCALE_H) || defined(HAVE_OS2_COUNTRYINFO) || defined(HAVE_WIN32_LOCALEINFO)
    cout << "  -C charset  Output charset     -L  Use locale's date "
            "format" << endl;
#else
    cout << "  -C charset  Output charset" << endl;
#endif
}

void evaluaterange(const char *rangespec,
                   time_t **range_start_pp, time_t **range_end_pp)
{
    // Allocate times if not already
    if (!*range_start_pp) *range_start_pp = new time_t;
    if (!*range_end_pp)   *range_end_pp = new time_t;

    // Evaluate range as given. It is on the form
    // "[yyyymmdd[Thhmmss]]-[yyyymmdd[Thhmmss]]"
    bool valid = false;
    const char *separator = strchr(rangespec, '-');
    if (separator)
    {
        if (0 == separator - rangespec)
        {
            **range_start_pp = time_t(0);
        }
        else
        {
            **range_start_pp =
                timespecToTimeT(string(rangespec, separator - rangespec));
        }

        if (time_t(-1) != **range_start_pp)
        {
            const char *rangespec2 = separator + 1;
            size_t charsleft = strlen(rangespec2);
            if (0 == charsleft)
            {
                **range_end_pp = time_t(DISTANT_FUTURE);
            }
            else
            {
                **range_end_pp =
                    timespecToTimeT(string(rangespec2, charsleft));
            }

            if (time_t(-1) != **range_end_pp)
            {
                valid = true;
            }
        }
    }

    if (!valid)
    {
        cerr << "Invalid range specification: " << rangespec << endl;
        exit(1);
    }
}
